src/util/JoinAlgorithms/JoinAlgorithms.h

```diff
@@ -497,6 +498,8 @@ void specialOptionalJoin(
     }
     it1 = next1;
 
+    checkCancellation();
```

Comment:
I think the one thing that is missing in this complete module is the cancellation checks in the `AddCombinedRowToTable` thing that is used to write the results. 
That way you can elegantly omit the timeout checks in the loops like 496 that might take some time during the materialization of the result.
And maybe you can add the cancelled resizing/reserving of the vectors there to figure out, if that was really our issue with the geometric stuff.

Reply:
This can be a different PR as it's relatively orthogonal to this thing here.